### ISSUE TITLE
Waiting for TaskCompletionSource blocks other requests

### ISSUE THREAD

**Issue by andreasjhkarlsson** [NancyFx/Nancy#2714]
When waiting for a TaskCompletionSource<T> in an async request handler it seems that all other requests are blocked until the TaskCompletionSource is completed.

Reproducer (tested with SelfHost):

```csharp
public class SampleModule : Nancy.NancyModule
{
    public SampleModule()
    {
        // Should return after 10 seconds
        Get("/threader", async (p)=>
        {
                
            var resultSource = new TaskCompletionSource<DateTime>();

            var thread = new Thread(
                () =>
                {
                    Thread.Sleep(10000);
                    resultSource.SetResult(DateTime.Now);
                }
            );
            thread.Start();

            var finished = await resultSource.Task;

            return string.Format("Thread finished at: {0}", finished);
        });


        // Should return right away
        Get("/echo", (p) =>
        {
            return "Echo?";
        });
    }
}
```

If you navigate to /threader in a browser tab and then open another tab and navigate to /echo, /echo won't respond until /threader completes also.

My guess is that it's because the wrapped Task never transitions to the 'Running' state, see: http://stackoverflow.com/questions/14836995/is-there-a-way-of-setting-a-task-driven-via-taskcompletionsource-to-status-runn

Tested with 2.0.0-clinteastwood, apologies if it's fixed in head.

**Comment by andreasjhkarlsson** [comment]
No, I just tried with a OwinHost and it does not seem to block other requests.

In fact, after some more testing it seems that all async requests blocks other requests when using Nancy.Hosting.Self. For example:

```csharp
public class SampleModule : Nancy.NancyModule
{
    public SampleModule()
    {
        // Should return after 10 seconds
        Get("/threader", async (p, ct) =>
        {
            await Task.Delay(10000);

            return string.Format("Thread finished at: {0}", DateTime.Now);
        });

        // Should return right away
        Get("/echo", (p) =>
        {
            return "Echo?";
        });
    }
}
```

So I guess that core issue is that async blocks on Nancy SelfHost.

**Comment by damianh** [comment]
Thank you for checking.

**Comment by khellang** [comment]
I think this might be because we're awaiting the request processing [here](https://github.com/NancyFx/Nancy/blob/ceea69506b9270c94cba231e0bf459897077058b/src/Nancy.Hosting.Self/NancyHost.cs#L134). See [this comment](https://github.com/NancyFx/Nancy/pull/2548#issuecomment-273596062) on the PR that introduced it. This is a pretty serious bug.

**Comment by xt0rted** [comment]
@khellang want me to send in a pr reverting that one line?

**Comment by khellang** [comment]
I think we should get #2697 in first :smile:

**Comment by davidallyoung** [comment]
I just got #2697 fixed from my git gaffe, and we should be able to have this resolved soon:tm: .

**Comment by khellang** [comment]
@andreasjhkarlsson Would you be able to pull the latest NuGet package from the MyGet feed and see if you can reproduce this now?

**Comment by andreasjhkarlsson** [comment]
2.0.0-Pre1846 still shows the same issue. Full reproducer solution here: https://minfil.org/l2sex6b4bd/reproducer.zip

**Comment by thecodejunkie** [comment]
ping @khellang @damianh

**Comment by damianh** [comment]
https://github.com/NancyFx/Nancy/blob/ceea69506b9270c94cba231e0bf459897077058b/src/Nancy.Hosting.Self/NancyHost.cs#L127-L142
^This looks fairly broken. It is handling one request at a time. It's not related to the `TaskCompletionSource` in OP's report _per se_.

https://github.com/aspnet/AspNetKatana/blob/dev/src/Microsoft.Owin.Host.HttpListener/OwinHttpListener.cs#L201-L202
^ Here you can see the MS Owin impl spawns a task for each request but it also does [reference counting](https://github.com/aspnet/AspNetKatana/blob/dev/src/Microsoft.Owin.Host.HttpListener/OwinHttpListener.cs#L210), "pump limits" etc.

I don't think I've ever used the `Nancy.Hosting.Self`; I've always used the MS Owin HttpListener host for my self hosting needs. (There is also OwinWebListenerHost). 

@NancyFx/most-valued-minions My opinion is to kill `Nancy.Hosting.Self` package. If it stays it needs a rewrite.

@andreasjhkarlsson In the mean time, use the MSOWIN host. V4 is on the way, so there is some legs left in it.



There is a v4 of owin http listener https://github.com/aspnet/AspNetKatana/milestones

**Comment by grumpydev** [comment]
This must have been broken fairlyish recently, it never used to only process one request at a time, but last time I looked at the code was probably pre-asyncawait updates.

**Comment by xt0rted** [comment]
This is from #2548 which is what I had asked about reverting in https://github.com/NancyFx/Nancy/issues/2714#issuecomment-285353026

**Comment by davidallyoung** [comment]
We actually changed Self Host earlier this year with this PR to resolve it only processing one request at a time with #2697.  I might be missing something here though :D

**Comment by cemremengu** [comment]
Kestrel is the way to go

**Comment by damianh** [comment]
Kestrel is nice, but the problem for use is that its public API is based on AspNet's HttpContext which is unfortunate.

**Comment by khellang** [comment]
@damianh https://github.com/khellang/KestrelPureOwin 😉

**Comment by damianh** [comment]
See the dependency graph? Not pure unfortunately.

**Comment by khellang** [comment]
That's nothing. Most of it isn't even used (the HTTP abstractions etc.). I bet you could trim some of the assemblies and it'd still run. And if you're targeting .NET Core 2.0, you already have it all on disk (it's not part of publish)

**Comment by jchannon** [comment]
Depends how you publish ie standalone app

**Comment by dpavsrtrl** [comment]
What's the path to fix this? Self-hosting is useful exactly because it has no dependencies.

**Comment by shmutalov** [comment]
Any news?

**Comment by youngbob** [comment]
This has already been fixed in #2697

**Comment by shmutalov** [comment]
Yep, just tested with latest builds. When it will be published to nuget.org?

**Comment by youngbob** [comment]
According to #2872, sometime this week or next.

**Comment by nkosi23** [comment]
Has this bug fix been released on NuGet?

**Comment by nkosi23** [comment]
Apparently not since the NancyHost options introduced in the commit are not available.

It would be really nice to have a snap 1.X release for Nancy.SelfHost since this is a major bug. The fix is already merged so hopefully releasing should be easy enough.

**Comment by nkosi23** [comment]
Pinging @thecodejunkie and @khellang

**Comment by nkosi23** [comment]
Just read in #2833 that apparently using the OWIN Self Host has been recommended "for a long time". Unfortunately I haven't seen this recommendation any in the documentation. We will try to switch to this SelfHost then.
